Compilation fails on osx

Compilation crashes in Basic.Gen.exe in native code. I'm running CppSharp for 64 bits on osx 10.10 using a recommended llvm x64 version. Is this crash known? Stacktrace:

```
==== Building CppSharp.AST (release_x64) ====
make[1]: Nothing to be done for `all'.
==== Building CppSharp.Runtime (release_x64) ====
make[1]: Nothing to be done for `all'.
==== Building CppSharp.CppParser (release_x64) ====
==== Building CppSharp.Parser.CSharp (release_x64) ====
make[1]: Nothing to be done for `all'.
==== Building CppSharp (release_x64) ====
make[1]: Nothing to be done for `all'.
==== Building CppSharp.Generator (release_x64) ====
make[1]: Nothing to be done for `all'.
==== Building CppSharp.Generator.Tests (release_x64) ====
make[1]: Nothing to be done for `all'.
==== Building Basic.Native (release_x64) ====
==== Building Basic.Gen (release_x64) ====
==== Building Basic.CSharp (release_x64) ====
Running pre-build commands
mono --debug "../lib/Release_x64/Basic.Gen.exe"
Stacktrace:

  at <unknown> <0xffffffff>
  at CppSharp.ConsoleDriver.Run (CppSharp.ILibrary) [0x00007] in /Users/Stephan/Sourcecode/CppSharp/src/Generator/Driver.cs:406
  at CppSharp.Tests.Basic.Main (string[]) [0x00007] in /Users/Stephan/Sourcecode/CppSharp/tests/Basic/Basic.cs:38
  at (wrapper runtime-invoke) <Module>.runtime_invoke_void_object (object,intptr,intptr,intptr) <IL 0x00050, 0xffffffff>

Native stacktrace:


Debug info from gdb:

(lldb) command source -s 0 '/tmp/mono-gdb-commands.aXTkz6'
Executing commands in '/private/tmp/mono-gdb-commands.aXTkz6'.
(lldb) process attach --pid 91290
Process 91290 stopped
Executable module set to "/usr/local/bin/mono".
Architecture set to: x86_64h-apple-macosx.
(lldb) thread list
Process 91290 stopped
* thread #1: tid = 0x436227, 0x00007fff8a798902 libsystem_kernel.dylib`__wait4 + 10, queue = 'com.apple.main-thread', stop reason = signal SIGSTOP
  thread #2: tid = 0x436228, 0x00007fff8a79351a libsystem_kernel.dylib`semaphore_wait_trap + 10
  thread #3: tid = 0x436229, 0x00007fff8a799232 libsystem_kernel.dylib`kevent64 + 10, queue = 'com.apple.libdispatch-manager'
(lldb) thread backtrace all
* thread #1: tid = 0x436227, 0x00007fff8a798902 libsystem_kernel.dylib`__wait4 + 10, queue = 'com.apple.main-thread', stop reason = signal SIGSTOP
  * frame #0: 0x00007fff8a798902 libsystem_kernel.dylib`__wait4 + 10
    frame #1: 0x000000010cb1d9d1 mono`mono_handle_native_sigsegv(signal=<unavailable>, ctx=<unavailable>) + 433 at mini-exceptions.c:2323
    frame #2: 0x000000010cb73e34 mono`mono_arch_handle_altstack_exception(sigctx=0x000000010cfecf48, fault_addr=<unavailable>, stack_ovf=0) + 84 at exceptions-amd64.c:861
    frame #3: 0x000000010ca83bab mono`mono_sigsegv_signal_handler(_dummy=<unavailable>, _info=<unavailable>, context=<unavailable>) + 379 at mini.c:6858
    frame #4: 0x00007fff8a25ef1a libsystem_platform.dylib`_sigtramp + 26
    frame #5: 0x000000010ccb0d51 mono`monoeg_g_hash_table_lookup [inlined] monoeg_g_hash_table_lookup_extended(hash=<unavailable>, value=<unavailable>) + 48 at ghashtable.c:331
    frame #6: 0x000000010ccb0d21 mono`monoeg_g_hash_table_lookup(hash=<unavailable>, key=0x0000000000002d6c) + 17 at ghashtable.c:311
    frame #7: 0x000000010cbba29e mono`mono_get_method_full(image=<unavailable>, token=<unavailable>, klass=<unavailable>, context=<unavailable>) + 158 at loader.c:1842
    frame #8: 0x000000010cbbc3be mono`find_method_in_class [inlined] mono_get_method(image=<unavailable>, token=<unavailable>, klass=0x00007fbeb38536e0) + 8 at loader.c:1821
    frame #9: 0x000000010cbbc3b6 mono`find_method_in_class(klass=0x00007fbeb38536e0, name=0x000000010f90aa5c, qname=0x0000000000000000, fqname=0x0000000000000000, sig=0x00007fbeb403ec70, from_class=0x00007fbeb38536e0) + 310 at loader.c:629
    frame #10: 0x000000010cbbc05f mono`find_method(in_class=<unavailable>, ic=<unavailable>, name=0x000000010f90aa5c, sig=0x00007fbeb403ec70, from_class=<unavailable>) + 271 at loader.c:699
    frame #11: 0x000000010cbbc859 mono`method_from_memberref(image=<unavailable>, idx=<unavailable>, typespec_context=<unavailable>, used_context=<unavailable>) + 697 at loader.c:1055
    frame #12: 0x000000010cbba42e mono`mono_get_method_from_token(image=0x00007fbeb4029200, token=<unavailable>, klass=<unavailable>, context=0x0000000000000000, used_context=<unavailable>) + 158 at loader.c:1726
    frame #13: 0x000000010cbba2c8 mono`mono_get_method_full(image=0x00007fbeb4029200, token=167772162, klass=0x0000000000000000, context=0x0000000000000000) + 200 at loader.c:1850
    frame #14: 0x000000010cc4898b mono`mono_custom_attrs_from_index(image=0x00007fbeb4029200, idx=46) + 331 at reflection.c:8748
    frame #15: 0x000000010ca8be0f mono`mono_method_to_ir [inlined] is_jit_optimizer_disabled(m=0x00007fbeb403d2b0) + 143 at method-to-ir.c:6762
    frame #16: 0x000000010ca8bd80 mono`mono_method_to_ir(cfg=0x00007fbeb400dc00, method=0x00007fbeb403d2b0, start_bblock=0x0000000000000000, end_bblock=0x0000000000000000, return_var=0x0000000000000000, inline_args=0x0000000000000000, inline_offset=<unavailable>, is_virtual_call=<unavailable>) + 64 at method-to-ir.c:7042
    frame #17: 0x000000010ca8060e mono`mini_method_compile(method=0x00007fbeb403d2b0, opts=370239999, domain=0x00007fbeb3414240, flags=<unavailable>, parts=0) + 2766 at mini.c:5281
    frame #18: 0x000000010ca82e3c mono`mono_jit_compile_method_with_opt [inlined] mono_jit_compile_method_inner(method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, method=0x00007fbeb403d2b0, opt=<unavailable>) + 194 at mini.c:6024
    frame #19: 0x000000010ca82d7a mono`mono_jit_compile_method_with_opt(method=<unavailable>, opt=<unavailable>, ex=0x00007fff53185fb0) + 1274 at mini.c:6296
    frame #20: 0x000000010ca8282a mono`mono_jit_compile_method(method=0x00007fbeb403d2b0) + 42 at mini.c:6333
    frame #21: 0x000000010cb1eeed mono`common_call_trampoline(m=<unavailable>, vtable_slot=<unavailable>, need_rgctx_tramp=0, regs=<unavailable>, code=<unavailable>, tramp=<unavailable>, vt=<unavailable>) + 1069 at mini-trampolines.c:584
    frame #22: 0x000000010cf91172
    frame #23: 0x000000010cfd314f
    frame #24: 0x000000010cfd2db4
    frame #25: 0x000000010ca86019 mono`mono_jit_runtime_invoke(method=<unavailable>, obj=0x0000000000000000, params=0x00007fff53186798, exc=0x000000010cfd2d70) + 1641 at mini.c:6735
    frame #26: 0x000000010cc2dd6e mono`mono_runtime_invoke(method=0x00007fbeb3415ba0, obj=0x0000000000000000, params=0x00007fff53186798, exc=0x0000000000000000) + 110 at object.c:2842
    frame #27: 0x000000010cc333a8 mono`mono_runtime_exec_main(method=0x00007fbeb3415ba0, args=<unavailable>, exc=0x0000000000000000) + 376 at object.c:4099
    frame #28: 0x000000010caef390 mono`mono_main [inlined] main_thread_handler + 7776 at driver.c:1064
    frame #29: 0x000000010caef354 mono`mono_main(argc=<unavailable>, argv=<unavailable>) + 7716 at driver.c:2022
    frame #30: 0x00007fff856e05c9 libdyld.dylib`start + 1
    frame #31: 0x00007fff856e05c9 libdyld.dylib`start + 1

  thread #2: tid = 0x436228, 0x00007fff8a79351a libsystem_kernel.dylib`semaphore_wait_trap + 10
    frame #0: 0x00007fff8a79351a libsystem_kernel.dylib`semaphore_wait_trap + 10
    frame #1: 0x000000010cca9187 mono`mono_sem_wait(sem=0x000000010cdcb908, alertable=1) + 23 at mono-semaphore.c:103
    frame #2: 0x000000010cc2b368 mono`finalizer_thread(unused=<unavailable>) + 120 at gc.c:1077
    frame #3: 0x000000010cc0b486 mono`start_wrapper [inlined] start_wrapper_internal + 426 at threads.c:663
    frame #4: 0x000000010cc0b2dc mono`start_wrapper(data=<unavailable>) + 44 at threads.c:710
    frame #5: 0x000000010ccae86e mono`inner_start_thread(arg=<unavailable>) + 222 at mono-threads-posix.c:88
    frame #6: 0x00007fff85893268 libsystem_pthread.dylib`_pthread_body + 131
    frame #7: 0x00007fff858931e5 libsystem_pthread.dylib`_pthread_start + 176
    frame #8: 0x00007fff8589141d libsystem_pthread.dylib`thread_start + 13

  thread #3: tid = 0x436229, 0x00007fff8a799232 libsystem_kernel.dylib`kevent64 + 10, queue = 'com.apple.libdispatch-manager'
    frame #0: 0x00007fff8a799232 libsystem_kernel.dylib`kevent64 + 10
    frame #1: 0x00007fff9241dd91 libdispatch.dylib`_dispatch_mgr_invoke + 247
    frame #2: 0x00007fff9241da6a libdispatch.dylib`_dispatch_mgr_thread + 52
(lldb) detach

=================================================================
Got a SIGSEGV while executing native code. This usually indicates
a fatal error in the mono runtime or one of the native libraries 
used by your application.
=================================================================

(lldb) quit
Process 91290 detached
(lldb) make[1]: *** [prebuild] Abort trap: 6
make: *** [Basic.CSharp] Error 2

```

I'm on a 64 bit mono and your fix did work :smile: 

The next problem is that a test is failing to compile:

```
Generating bindings for CSharpTemp (CSharp)
Looking for tests in: /Users/Stephan/Sourcecode/CppSharp/tests/CSharpTemp
Parsed 'AnotherUnit.h'
Error parsing 'CSharpTemp.h'
/Users/Stephan/Sourcecode/CppSharp/tests/CSharpTemp/CSharpTemp.h(2,10): fatal: 'cstdint' file not found
Generated 'CSharpTemp.cs'
Generating response file
==== Building CSharpTemp.Tests.CSharp (release_x64) ====
Creating ../obj/x64/Release/CSharpTemp.Tests.CSharp
Generating response file
../../../tests/CSharpTemp/CSharpTemp.Tests.cs(6,13): error CS0234: The type or namespace name `Foo' does not exist in the namespace `CSharpTemp'. Are you missing an assembly reference?
Compilation failed: 1 error(s), 0 warnings
make[1]: *** [../lib/Release_x64/CSharpTemp.Tests.CSharp.dll] Error 1
make: *** [CSharpTemp.Tests.CSharp] Error 2
```
It seems like something is not included during the build, I'm investigating

It seems like that's really a similar issue.
I enabled verbose logging the include path for the compiled clang is in there but there are more pointing to the clang installed in /usr. Maybe something gets confused between Apple's clang in /usr and the self compiled one. Do I need to install the compiled clang? Here's the verbose log:

```
Generating bindings for CSharpTemp (CSharp)
Looking for tests in: /Users/Stephan/Sourcecode/CppSharp/tests/CSharpTemp
Parsing libraries...
Indexing library symbols...
Parsing code...
ignoring nonexistent directory "./lib/clang/3.7.0/include"
ignoring nonexistent directory "/usr/include/c++/4.2.1/i686-apple-darwin10/x86_64"
ignoring nonexistent directory "/usr/include/c++/4.0.0"
ignoring nonexistent directory "/usr/include/c++/4.0.0/i686-apple-darwin8/"
ignoring nonexistent directory "/usr/include/c++/4.0.0/backward"
ignoring nonexistent directory "./lib/clang/3.7.0/include"
#include "..." search starts here:
#include <...> search starts here:
 /Users/Stephan/Sourcecode/CppSharp/tests/CSharpTemp
 /Users/Stephan/Sourcecode/CppSharp/deps/llvm/tools/clang/lib/Headers
 /usr/include/c++/4.2.1
 /usr/include/c++/4.2.1/backward
 /usr/local/include
 /usr/include
 /System/Library/Frameworks (framework directory)
 /Library/Frameworks (framework directory)
End of search list.
Parsed 'AnotherUnit.h'
ignoring nonexistent directory "./lib/clang/3.7.0/include"
ignoring nonexistent directory "/usr/include/c++/4.2.1/i686-apple-darwin10/x86_64"
ignoring nonexistent directory "/usr/include/c++/4.0.0"
ignoring nonexistent directory "/usr/include/c++/4.0.0/i686-apple-darwin8/"
ignoring nonexistent directory "/usr/include/c++/4.0.0/backward"
ignoring nonexistent directory "./lib/clang/3.7.0/include"
#include "..." search starts here:
#include <...> search starts here:
 /Users/Stephan/Sourcecode/CppSharp/tests/CSharpTemp
 /Users/Stephan/Sourcecode/CppSharp/deps/llvm/tools/clang/lib/Headers
 /usr/include/c++/4.2.1
 /usr/include/c++/4.2.1/backward
 /usr/local/include
 /usr/include
 /System/Library/Frameworks (framework directory)
 /Library/Frameworks (framework directory)
End of search list.
Error parsing 'CSharpTemp.h'
/Users/Stephan/Sourcecode/CppSharp/tests/CSharpTemp/CSharpTemp.h(2,10): fatal: 'cstdint' file not found
ignoring nonexistent directory "./lib/clang/3.7.0/include"
ignoring nonexistent directory "/usr/include/c++/4.2.1/i686-apple-darwin10/x86_64"
ignoring nonexistent directory "/usr/include/c++/4.0.0"
ignoring nonexistent directory "/usr/include/c++/4.0.0/i686-apple-darwin8/"
ignoring nonexistent directory "/usr/include/c++/4.0.0/backward"
ignoring nonexistent directory "./lib/clang/3.7.0/include"
#include "..." search starts here:
#include <...> search starts here:
 /Users/Stephan/Sourcecode/CppSharp/tests/CSharpTemp
 /Users/Stephan/Sourcecode/CppSharp/deps/llvm/tools/clang/lib/Headers
 /usr/include/c++/4.2.1
 /usr/include/c++/4.2.1/backward
 /usr/local/include
 /usr/include
 /System/Library/Frameworks (framework directory)
 /Library/Frameworks (framework directory)
End of search list.
Processing code...
Pass 'CppSharp.Passes.CleanUnitPass'
Pass 'CppSharp.Passes.SortDeclarationsPass'
Pass 'CppSharp.Passes.ResolveIncompleteDeclsPass'
Pass 'CppSharp.Passes.CheckIgnoredDeclsPass'
Pass 'CppSharp.Tests.TestAttributesPass'
Pass 'CppSharp.Passes.FindSymbolsPass'
Pass 'CppSharp.Passes.CheckStaticClass'
Pass 'CppSharp.Passes.MoveOperatorToClassPass'
Pass 'CppSharp.Passes.MoveFunctionToClassPass'
Pass 'CppSharp.Passes.ConstructorToConversionOperatorPass'
Pass 'CppSharp.Passes.CheckAmbiguousFunctions'
Pass 'CppSharp.Passes.CheckOperatorsOverloadsPass'
Pass 'CppSharp.Passes.CheckVirtualOverrideReturnCovariance'
Pass 'CppSharp.Passes.CheckAbiParameters'
Pass 'CppSharp.Passes.FieldToPropertyPass'
Pass 'CppSharp.Passes.CleanInvalidDeclNamesPass'
Pass 'CppSharp.Passes.CheckIgnoredDeclsPass'
Pass 'CppSharp.Passes.CheckFlagEnumsPass'
Pass 'CppSharp.Passes.CheckDuplicatedNamesPass'
Pass 'CppSharp.Passes.HandleDefaultParamValuesPass'
Pass 'CppSharp.Passes.FixDefaultParamValuesOfOverridesPass'
Pass 'CppSharp.Passes.GenerateAbstractImplementationsPass'
Pass 'CppSharp.Passes.MultipleInheritancePass'
Pass 'CppSharp.Passes.ParamTypeToInterfacePass'
Pass 'CppSharp.Passes.CheckVTableComponentsPass'
Pass 'CppSharp.Passes.GetterSetterToPropertyAdvancedPass'
Pass 'CppSharp.Passes.RenameRootNamespacesPass'
Generating code...
Generated 'CSharpTemp.cs'
==== Building CSharpTemp.Tests.CSharp (release_x64) ====
../../../tests/CSharpTemp/CSharpTemp.Tests.cs(6,13): error CS0234: The type or namespace name `Foo' does not exist in the namespace `CSharpTemp'. Are you missing an assembly reference?
Compilation failed: 1 error(s), 0 warnings
make[1]: *** [../lib/Release_x64/CSharpTemp.Tests.CSharp.dll] Error 1
make: *** [CSharpTemp.Tests.CSharp] Error 2
```

--- a/src/Generator.Tests/GeneratorTest.cs
+++ b/src/Generator.Tests/GeneratorTest.cs
@@ -49,6 +49,10 @@ namespace CppSharp.Utils
                 Path.GetFullPath(Path.Combine(path, "../../deps/llvm/tools/clang/lib/Headers"))
             };
 
+            if (IsMacOS) {
+                options.addArguments ("-stdlib=libc++");
+            }
+
             foreach (var header in headersPaths)
                 options.addSystemIncludeDirs(header);
 
@@ -104,6 +108,28 @@ namespace CppSharp.Utils
             }
 
             throw new Exception("Could not find tests output directory");
+        }
+
+        [DllImport ("libc")]
+        static extern int uname (IntPtr buf);
+
+        public static bool IsMacOS {
+            get {
+                if (Environment.OSVersion.Platform != PlatformID.Unix)
+                    return false;
+
+                IntPtr buf = Marshal.AllocHGlobal (8192);
+                if (uname (buf) == 0) {
+                    string os = Marshal.PtrToStringAnsi (buf);
+                    switch (os) {
+                    case "Darwin":
+                        return true;
+                    }
+                }
+                Marshal.FreeHGlobal (buf);
+
+                return false;
+            }
         }
         #endregion
     }

